doc/pcreposix.3
.TH PCREPOSIX 3 "09 January 2012" "PCRE 8.30"
.SH NAME
PCRE - Perl-compatible regular expressions.
.SH "SYNOPSIS"
.rs
.sp
.B #include <pcreposix/pcreposix.h>
.PP
.nf
.B int regcomp(regex_t *\fIpreg\fP, const char *\fIpattern\fP,
.B "     int \fIcflags\fP);"
.sp
.B int regexec(regex_t *\fIpreg\fP, const char *\fIstring\fP,
.B "     size_t \fInmatch\fP, regmatch_t \fIpmatch\fP[], int \fIeflags\fP);"
.B "     size_t regerror(int \fIerrcode\fP, const regex_t *\fIpreg\fP,"
.B "     char *\fIerrbuf\fP, size_t \fIerrbuf_size\fP);"
.sp
.B void regfree(regex_t *\fIpreg\fP);
.fi
.
.SH DESCRIPTION
.rs
.sp
This man page reflects the old version of the pcreposix API and library.
To get the behavior described here with the new pcreposix API and library
the "REG_PCRE" option must be passed to the "regcomp" function.  See the
\fBpcreposix/pcreposix.h\fP header file for full details.
.P
This set of functions provides a POSIX-style API for the PCRE regular
expression 8-bit library. See the
.\" HREF
\fBpcreapi\fP
.\"
documentation for a description of PCRE's native API, which contains much
additional functionality. There is no POSIX-style wrapper for PCRE's 16-bit
and 32-bit library.
.P
The functions described here are just wrapper functions that ultimately call
the PCRE native API. Their prototypes are defined in the \fBpcreposix/pcreposix.h\fP
header file, and on Unix systems the library itself is called
\fBpcreposix.a\fP, so can be accessed by adding \fB-lpcreposix\fP to the
command for linking an application that uses them. Because the POSIX functions
call the native ones, it is also necessary to add \fB-lpcre\fP.
.P
I have implemented only those POSIX option bits that can be reasonably mapped
to PCRE native options. In addition, the option REG_EXTENDED is defined with
the value zero. This has no effect, but since programs that are written to the
POSIX interface often use it, this makes it easier to slot in PCRE as a
replacement library. Other POSIX options are not even defined.
.P
There are also some other options that are not defined by POSIX. These have
been added at the request of users who want to make use of certain
PCRE-specific features via the POSIX calling interface.
.P
When PCRE is called via these functions, it is only the API that is POSIX-like
in style. The syntax and semantics of the regular expressions themselves are
still those of Perl, subject to the setting of various PCRE options, as
described below. "POSIX-like in style" means that the API approximates to the
POSIX definition; it is not fully POSIX-compatible, and in multi-byte encoding
domains it is probably even less compatible.
.P
The header for these functions is supplied as \fBpcreposix/pcreposix.h\fP to avoid any
potential clash with other POSIX libraries. It can, of course, be renamed or
aliased as \fBregex.h\fP, which is the "correct" name. It provides two
structure types, \fIregex_t\fP for compiled internal forms, and
\fIregmatch_t\fP for returning captured substrings. It also defines some
constants whose names start with "REG_"; these are used for setting options and
identifying error codes.
.
.
.SH "COMPILING A PATTERN"
.rs
.sp
The function \fBregcomp()\fP is called to compile a pattern into an
internal form. The pattern is a C string terminated by a binary zero, and
is passed in the argument \fIpattern\fP. The \fIpreg\fP argument is a pointer
to a \fBregex_t\fP structure that is used as a base for storing information
about the compiled regular expression.
.P
The argument \fIcflags\fP is either zero, or contains one or more of the bits
defined by the following macros:
.sp
  REG_DOTALL
.sp
The PCRE_DOTALL option is set when the regular expression is passed for
compilation to the native function. Note that REG_DOTALL is not part of the
POSIX standard.
.sp
  REG_ICASE
.sp
The PCRE_CASELESS option is set when the regular expression is passed for
compilation to the native function.
.sp
  REG_NEWLINE
.sp
When this option is \fInot\fP set both PCRE_DOTALL and PCRE_DOLLAR_ENDONLY
are always passed to the native function.  When this option \fIis\fP set
PCRE_MULTILINE and PCRE_NOT_EXCLUDES_NL are always passed to the native
function. Note that this \fIdoes\fP mimic the defined POSIX behaviour for
REG_NEWLINE (see the following section).
.sp
  REG_NOSUB
.sp
The PCRE_NO_AUTO_CAPTURE option is set when the regular expression is passed
for compilation to the native function. In addition, when a pattern that is
compiled with this flag is passed to \fBregexec()\fP for matching, the
\fInmatch\fP and \fIpmatch\fP arguments are ignored, and no captured strings
are returned.
.sp
  REG_UCP
.sp
The PCRE_UCP option is set when the regular expression is passed for
compilation to the native function. This causes PCRE to use Unicode properties
when matchine \ed, \ew, etc., instead of just recognizing ASCII values. Note
that REG_UTF8 is not part of the POSIX standard.
.sp
  REG_UNGREEDY
.sp
The PCRE_UNGREEDY option is set when the regular expression is passed for
compilation to the native function. Note that REG_UNGREEDY is not part of the
POSIX standard.
.sp
  REG_UTF8
.sp
The PCRE_UTF8 option is set when the regular expression is passed for
compilation to the native function. This causes the pattern itself and all data
strings used for matching it to be treated as UTF-8 strings. Note that REG_UTF8
is not part of the POSIX standard.
.sp
  REG_MULTILINE
.sp
The PCRE_MULTILINE option is set when the regular expression is passed for
compilation to the native function. Note that this does \fInot\fP mimic the
defined POSIX behaviour for REG_NEWLINE (see the following section).
.P
In the absence of these flags, PCRE_DOTALL and PCRE_DOLLAR_ENDONLY are passed
to the native function. This means the regex is compiled with POSIX default
semantics. In particular, the way it handles newline characters in the subject
string is the POSIX way, not the Perl way. Note that setting REG_MULTILINE has
only \fIsome\fP of the effects specified for REG_NEWLINE. It does not affect
the way newlines are matched by . (they are not) or by a negative class such as
[^a] (they are).  However, setting REG_NEWLINE suppresses automatically passing
PCRE_DOTALL (but REG_DOTALL can be used to set it) and PCRE_DOLLAR_ENDONLY and
instead causes PCRE_MULTILINE (same as REG_MULTILINE) and PCRE_NOT_EXCLUDES_NL
to be passed to the native function which mimics POSIX REG_NEWLINE behaviour.
.P
The yield of \fBregcomp()\fP is zero on success, and non-zero otherwise. The
\fIpreg\fP structure is filled in on success, and one member of the structure
is public: \fIre_nsub\fP contains the number of capturing subpatterns in
the regular expression. Various error codes are defined in the header file.
.P
NOTE: If the yield of \fBregcomp()\fP is non-zero, you must not attempt to
use the contents of the \fIpreg\fP structure. If, for example, you pass it to
\fBregexec()\fP, the result is undefined and your program is likely to crash.
.
.
.SH "MATCHING NEWLINE CHARACTERS"
.rs
.sp
This area is not simple, because POSIX and Perl take different views of things.
It is possible to get PCRE to mostly obey POSIX semantics and that is what
pcreposix attempts to do. The following table lists the different
possibilities for matching newline characters in PCRE:
.sp
                          Default   Change with
.sp
  . matches newline          no     PCRE_DOTALL
  newline matches [^a]       yes    PCRE_NOT_EXCLUDES_NL
  $ matches \en at end        yes    PCRE_DOLLAR_ENDONLY
  $ matches \en in middle     no     PCRE_MULTILINE
  ^ matches \en in middle     no     PCRE_MULTILINE
.sp
This is the equivalent table for POSIX:
.sp
                          Default   Change with
.sp
  . matches newline          yes    REG_NEWLINE
  newline matches [^a]       yes    REG_NEWLINE
  $ matches \en at end        no     REG_NEWLINE
  $ matches \en in middle     no     REG_NEWLINE
  ^ matches \en in middle     no     REG_NEWLINE
.sp
PCRE's behaviour is the same as Perl's, except that there is no equivalent for
PCRE_DOLLAR_ENDONLY in Perl. In Perl there is no way to stop newline from
matching [^a] like PCRE's PCRE_NOT_EXCLUDES_NL option does.
.P
The default POSIX newline handling can be obtained by setting PCRE_DOTALL and
PCRE_DOLLAR_ENDONLY which is what happens when REG_NEWLINE is NOT set. When
REG_NEWLINE IS set PCRE_NOT_EXCLUDES_NL and PCRE_MULTILINE are set instead to
mimic the POSIX behavior exactly in that case.
.
.
.SH "MATCHING A PATTERN"
.rs
.sp
The function \fBregexec()\fP is called to match a compiled pattern \fIpreg\fP
against a given \fIstring\fP, which is by default terminated by a zero byte
(but see REG_STARTEND below), subject to the options in \fIeflags\fP. These can
be:
.sp
  REG_NOTBOL
.sp
The PCRE_NOTBOL option is set when calling the underlying PCRE matching
function.
.sp
  REG_NOTEMPTY
.sp
The PCRE_NOTEMPTY option is set when calling the underlying PCRE matching
function. Note that REG_NOTEMPTY is not part of the POSIX standard. However,
setting this option can give more POSIX-like behaviour in some situations.
.sp
  REG_NOTEOL
.sp
The PCRE_NOTEOL option is set when calling the underlying PCRE matching
function.
.sp
  REG_STARTEND
.sp
The string is considered to start at \fIstring\fP + \fIpmatch[0].rm_so\fP and
to have a terminating NUL located at \fIstring\fP + \fIpmatch[0].rm_eo\fP

pcreposix.c
/*************************************************
*      Perl-Compatible Regular Expressions       *
*************************************************/

/* PCRE is a library of functions to support regular expressions whose syntax
and semantics are as close as possible to those of the Perl 5 language.

                       Written by Philip Hazel
           Copyright (c) 1997-2020 University of Cambridge
           Copyright (C) 2017-2020 Kyle J. McKay <mackyle@gmail.com>
           All Rights Reserved

-----------------------------------------------------------------------------
Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright notice,
      this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.

    * Neither the name of the University of Cambridge nor the names of its
      contributors may be used to endorse or promote products derived from
      this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.
-----------------------------------------------------------------------------
*/


/* This module is a wrapper that provides a POSIX API to the underlying PCRE
functions. */


#ifdef HAVE_CONFIG_H
#include "config.h"
#endif


/* Ensure that the PCREPOSIX_EXP_xxx macros are set appropriately for
compiling these functions. This must come before including pcreposix.h, where
they are set for an application (using these functions) if they have not
previously been set. */

#if defined(_WIN32) && !defined(PCRE_STATIC)
#  define PCREPOSIX_EXP_DECL extern __declspec(dllexport)
#  define PCREPOSIX_EXP_DEFN __declspec(dllexport)
#endif

/* We include pcre.h before pcre_internal.h so that the PCRE library functions
are declared as "import" for Windows by defining PCRE_EXP_DECL as "import".
This is needed even though pcre_internal.h itself includes pcre.h, because it
does so after it has set PCRE_EXP_DECL to "export" if it is not already set. */

#include "pcre.h"
#include "pcre_internal.h"
#include "pcreposix.h"

/* Known option bits */

#define REGCOMP_OPTIONS \
  (REG_ICASE|REG_NEWLINE|REG_DOTALL|REG_NOSUB|REG_UTF8|REG_PEND|REG_UNGREEDY| \
   REG_UCP|REG_MULTILINE|REG_NOSPEC|REG_EXPANDED|REG_BASIC|REG_EXTENDED| \
   REG_PCRE|REG_DENDONLY|REG_ANCHORED|REG_JAVASCPT)

#define REGEXEC_OPTIONS \
  (REG_NOTBOL|REG_NOTEOL|REG_STARTEND|REG_NOTEMPTY)


/* Table to translate PCRE compile time error codes into POSIX error codes. */

static const int eint[] = {
  0,           /* no error */
  REG_EESCAPE, /* \ at end of pattern */
  REG_EESCAPE, /* \c at end of pattern */
  REG_EESCAPE, /* unrecognized character follows \ */
  REG_BADBR,   /* numbers out of order in {} quantifier */
  /* 5 */
  REG_BADBR,   /* number too big in {} quantifier */
  REG_EBRACK,  /* missing terminating ] for character class */
  REG_ECTYPE,  /* invalid escape sequence in character class */
  REG_ERANGE,  /* range out of order in character class */
  REG_BADRPT,  /* nothing to repeat */
  /* 10 */
  REG_BADRPT,  /* operand of unlimited repeat could match the empty string */
  REG_ASSERT,  /* internal error: unexpected repeat */
  REG_BADPAT,  /* unrecognized character after (? */
  REG_BADPAT,  /* POSIX named classes are supported only within a class */
  REG_EPAREN,  /* missing ) */
  /* 15 */
  REG_ESUBREG, /* reference to non-existent subpattern */
  REG_INVARG,  /* erroffset passed as NULL */
  REG_INVARG,  /* unknown option bit(s) set */
  REG_EPAREN,  /* missing ) after comment */
  REG_ESIZE,   /* parentheses nested too deeply */
  /* 20 */
  REG_ESIZE,   /* regular expression too large */
  REG_ESPACE,  /* failed to get memory */
  REG_EPAREN,  /* unmatched parentheses */
  REG_ASSERT,  /* internal error: code overflow */
  REG_BADPAT,  /* unrecognized character after (?< */
  /* 25 */
  REG_BADPAT,  /* lookbehind assertion is not fixed length */
  REG_BADPAT,  /* malformed number or name after (?( */
  REG_BADPAT,  /* conditional group contains more than two branches */
  REG_BADPAT,  /* assertion expected after (?( */
  REG_BADPAT,  /* (?R or (?[+-]digits must be followed by ) */
  /* 30 */
  REG_ECTYPE,  /* unknown POSIX class name */
  REG_BADPAT,  /* POSIX collating elements are not supported */
  REG_INVARG,  /* this version of PCRE is not compiled with PCRE_UTF8 support */
  REG_BADPAT,  /* spare error */
  REG_BADPAT,  /* character value in \x{} or \o{} is too large */
  /* 35 */
  REG_BADPAT,  /* invalid condition (?(0) */
  REG_BADPAT,  /* \C not allowed in lookbehind assertion */
  REG_EESCAPE, /* PCRE does not support \L, \l, \N, \U, or \u */
  REG_BADPAT,  /* number after (?C is > 255 */
  REG_BADPAT,  /* closing ) for (?C expected */
  /* 40 */
  REG_BADPAT,  /* recursive call could loop indefinitely */
  REG_BADPAT,  /* unrecognized character after (?P */
  REG_BADPAT,  /* syntax error in subpattern name (missing terminator) */
  REG_BADPAT,  /* two named subpatterns have the same name */
  REG_BADPAT,  /* invalid UTF-8 string */
  /* 45 */
  REG_BADPAT,  /* support for \P, \p, and \X has not been compiled */
  REG_BADPAT,  /* malformed \P or \p sequence */
  REG_BADPAT,  /* unknown property name after \P or \p */
  REG_BADPAT,  /* subpattern name is too long (maximum 32 characters) */
  REG_BADPAT,  /* too many named subpatterns (maximum 10,000) */
  /* 50 */
  REG_BADPAT,  /* repeated subpattern is too long */
  REG_BADPAT,  /* octal value is greater than \377 (not in UTF-8 mode) */
  REG_BADPAT,  /* internal error: overran compiling workspace */
  REG_BADPAT,  /* internal error: previously-checked referenced subpattern not found */
  REG_BADPAT,  /* DEFINE group contains more than one branch */
  /* 55 */
  REG_BADPAT,  /* repeating a DEFINE group is not allowed */
  REG_INVARG,  /* inconsistent NEWLINE options */
  REG_BADPAT,  /* \g is not followed followed by an (optionally braced) non-zero number */
  REG_BADPAT,  /* a numbered reference must not be zero */
  REG_BADPAT,  /* an argument is not allowed for (*ACCEPT), (*FAIL), or (*COMMIT) */
  /* 60 */
  REG_BADPAT,  /* (*VERB) not recognized */
  REG_BADPAT,  /* number is too big */
  REG_BADPAT,  /* subpattern name expected */
  REG_BADPAT,  /* digit expected after (?+ */
  REG_BADPAT,  /* ] is an invalid data character in JavaScript compatibility mode */
  /* 65 */
  REG_BADPAT,  /* different names for subpatterns of the same number are not allowed */
  REG_BADPAT,  /* (*MARK) must have an argument */
  REG_INVARG,  /* this version of PCRE is not compiled with PCRE_UCP support */
  REG_BADPAT,  /* \c must be followed by an ASCII character */
  REG_BADPAT,  /* \k is not followed by a braced, angle-bracketed, or quoted name */
  /* 70 */
  REG_BADPAT,  /* internal error: unknown opcode in find_fixedlength() */
  REG_BADPAT,  /* \N is not supported in a class */
  REG_BADPAT,  /* too many forward references */
  REG_BADPAT,  /* disallowed UTF-8/16/32 code point (>= 0xd800 && <= 0xdfff) */
  REG_BADPAT,  /* invalid UTF-16 string (should not occur) */
  /* 75 */
  REG_BADPAT,  /* overlong MARK name */
  REG_BADPAT,  /* character value in \u.... sequence is too large */
  REG_BADPAT,  /* invalid UTF-32 string (should not occur) */
  REG_BADPAT,  /* setting UTF is disabled by the application */
  REG_BADPAT,  /* non-hex character in \\x{} (closing brace missing?) */
  /* 80 */
  REG_BADPAT,  /* non-octal character in \o{} (closing brace missing?) */
  REG_BADPAT,  /* missing opening brace after \o */
  REG_BADPAT,  /* parentheses too deeply nested */
  REG_BADPAT,  /* invalid range in character class */
  REG_BADPAT,  /* group name must start with a non-digit */
  /* 85 */
  REG_BADPAT,  /* parentheses too deeply nested (stack check) */
  REG_BADPAT,  /* missing digits in \x{} or \o{} */
  REG_BADPAT   /* pattern too complicated */
};

/* Table of texts corresponding to POSIX error codes */

static const char *const pstring[] = {
  "",                                /* Dummy for value 0 */
  "internal error",                  /* REG_ASSERT */
  "invalid repeat counts in {}",     /* BADBR      */
  "pattern error",                   /* BADPAT     */
  "? * + invalid",                   /* BADRPT     */
  "unbalanced {}",                   /* EBRACE     */
  "unbalanced []",                   /* EBRACK     */
  "collation error - not relevant",  /* ECOLLATE   */
  "bad class",                       /* ECTYPE     */
  "bad escape sequence",             /* EESCAPE    */
  "empty expression",                /* EMPTY      */
  "unbalanced ()",                   /* EPAREN     */
  "bad range inside []",             /* ERANGE     */
  "expression too big",              /* ESIZE      */
  "failed to get memory",            /* ESPACE     */
  "bad back reference",              /* ESUBREG    */
  "bad argument",                    /* INVARG     */
  "match failed"                     /* NOMATCH    */
};

#define PCREPOSIX_JIT_MAGIC 0x50435258UL /* PCRX */
#define GET_MAGIC(p) (((const pcrx_t *)(p))->magic_number)

/* internal regex_t re_impl struct */
typedef struct {
  pcre_uint32 magic_number;
  pcre_uint32 size;
  pcre *re_pcre;
  pcre_extra *re_pcre_extra;
} pcrx_t;




/*************************************************
*          Translate error code to string        *
*************************************************/

PCREPOSIX_EXP_DEFN size_t PCRE_CALL_CONVENTION
regerror(int errcode, const regex_t *preg, char *errbuf, size_t errbuf_size)
{
const char *message, *addmessage;
size_t length, addlength;

message = (errcode >= (int)(sizeof(pstring)/sizeof(char *)))?
  "unknown error code" : pstring[errcode];
length = strlen(message) + 1;

addmessage = " at offset ";
addlength = (preg != NULL && (int)preg->re_erroffset != -1)?
  strlen(addmessage) + 6 : 0;

if (errbuf_size > 0)
  {
  if (addlength > 0 && errbuf_size >= length + addlength)
    sprintf(errbuf, "%s%s%-6d", message, addmessage, (int)preg->re_erroffset);
  else
    {
    strncpy(errbuf, message, errbuf_size - 1);
    errbuf[errbuf_size-1] = 0;
    }
  }

return length + addlength;
}




/*************************************************
*           Free store held by a regex           *
*************************************************/

PCREPOSIX_EXP_DEFN void PCRE_CALL_CONVENTION
regfree(regex_t *preg)
{
if (preg == NULL || preg->re_impl == NULL)
  return;
if (GET_MAGIC(preg->re_impl) == PCREPOSIX_JIT_MAGIC)
  {
  pcrx_t *pjit = (pcrx_t *)preg->re_impl;
  if (pjit->re_pcre_extra)
    pcre_free_study(pjit->re_pcre_extra);
  preg->re_impl = pjit->re_pcre;
  (PUBL(free))(pjit);
  }
if (preg->re_impl != NULL)
  (PUBL(free))(preg->re_impl);
}




/*************************************************
*      Jitify a compiled regular expression      *
*************************************************/

/*
Arguments:
  preg        points to a structure with the compiled expression

Returns:      adds JIT compilation data if supported
*/

static void
regcomp_jit(regex_t *preg)
{
pcrx_t *pjit;
pcre *re_pcre;
const char *errorptr;
int hasjit = 0;

if (!preg || !preg->re_impl ||
    pcre_config(PCRE_CONFIG_JIT, &hasjit) != 0 || !hasjit)
  return;

pjit = (pcrx_t *)(PUBL(malloc))(sizeof(pcrx_t));
if (!pjit)
  return;
re_pcre = (pcre *)preg->re_impl;
hasjit = 0;
errorptr = NULL;
pjit->re_pcre_extra = NULL;
pjit->re_pcre_extra = pcre_study(re_pcre, PCRE_STUDY_JIT_COMPILE, &errorptr);
if (!pjit->re_pcre_extra || errorptr != NULL ||
    pcre_fullinfo(re_pcre, pjit->re_pcre_extra, PCRE_INFO_JIT, &hasjit) != 0 ||
    !hasjit)
  {
  if (pjit->re_pcre_extra)
    pcre_free_study(pjit->re_pcre_extra);
  (PUBL(free))(pjit);
  return;
  }
pjit->size = (pcre_uint32)-1L;
pjit->re_pcre = re_pcre;
pjit->magic_number = PCREPOSIX_JIT_MAGIC;
preg->re_impl = pjit;
}




/*************************************************
*            Compile a regular expression        *
*************************************************/

/*
Arguments:
  preg        points to a structure for recording the compiled expression
  pattern     the pattern to compile
  cflags      compilation flags

Returns:      0 on success
              various non-zero codes on failure
*/

PCREPOSIX_EXP_DEFN int PCRE_CALL_CONVENTION
regcomp(regex_t *preg, const char *pattern, int cflags)
{
const char *errorptr;
char *dupez = NULL;
int erroffset;
int errorcode;
int options = 0;
int re_nsub = 0;
const char *vptrs[4];
char altpattern[(sizeof(void *) * 2) + 1];

if (preg == NULL) return REG_INVARG;
preg->re_impl = NULL;
if ((cflags & REGCOMP_OPTIONS) != cflags) return REG_INVARG;

if ((cflags & REG_PEND) != 0)
  {
  uintptr_t page1, page2, ptr;
  int i;

  if (preg->re_endp < pattern) return REG_INVARG;
  if (pattern == preg->re_endp)
    {
    pattern = "";
    cflags &= ~REG_PEND;
    goto CHECK_OPTS;
    }
  if (pattern == NULL) return REG_INVARG;
# define PAGE1KM1 ((uintptr_t)(1024U-1U))
  page2 = (uintptr_t)(void *)preg->re_endp;
  page1 = page2 - 1U;
  page1 &= ~PAGE1KM1;
  page2 &= ~PAGE1KM1;
# undef PAGE1KM1
  if (page1 == page2 && *(preg->re_endp) == 0)
    {
    vptrs[2] = pattern;
    vptrs[3] = preg->re_endp;
    goto BUILD_PTRS;
    }
  dupez = (char *)malloc((preg->re_endp - pattern) + 1);
  if (!dupez) return REG_ESPACE;
  vptrs[2] = dupez;
  vptrs[3] = dupez + (preg->re_endp - pattern);
  memcpy(dupez, pattern, preg->re_endp - pattern);
  dupez[preg->re_endp - pattern] = 0;
  BUILD_PTRS:
# define XPTR(x) ((const char *)(void *)(uintptr_t)(x))
  vptrs[0] = XPTR(0x4841434b /* odd magic number ;) */);
  vptrs[1] = XPTR(0x01010101 /* version number one */);
# undef XPTR
  ptr = (uintptr_t)(void *)&vptrs[0];
  i = sizeof(void *) * 2;
  altpattern[i] = 0;
  while (i--)
    {
    unsigned nibble = ptr & 0xf;
    ptr >>= 4;
    altpattern[i] = (nibble <= 9) ? (CHAR_0 + nibble) : (CHAR_a + (nibble - 10));
    }
  pattern = altpattern;
  }
else if (pattern == NULL) return REG_INVARG;

CHECK_OPTS:
#define USE_EXTENDED (REG_EXTENDED|REG_PCRE|REG_JAVASCPT)
if ((cflags & USE_EXTENDED) == 0) options |= PCRE_POSIX_BASIC_ESC;
if ((cflags & USE_EXTENDED) == REG_EXTENDED) options |= PCRE_POSIX_EXTENDED;

if ((cflags & REG_PEND) != 0)     options |= PCRE_ALLOW_EMBEDDED_NUL;

if ((cflags & REG_ICASE) != 0)    options |= PCRE_CASELESS;
#define REG_XXXLINE (REG_MULTILINE|REG_NEWLINE)
if ((cflags & REG_XXXLINE) != 0)  options |= PCRE_MULTILINE;
if ((cflags & REG_DOTALL) != 0)   options |= PCRE_DOTALL;
if ((cflags & REG_NOSUB) != 0)    options |= PCRE_NO_AUTO_CAPTURE;
if ((cflags & REG_UTF8) != 0)     options |= PCRE_UTF8;
if ((cflags & REG_UCP) != 0)      options |= PCRE_UCP;
if ((cflags & REG_UNGREEDY) != 0) options |= PCRE_UNGREEDY;
if ((cflags & REG_NOSPEC) != 0)   options |= PCRE_VERBATIM;
if ((cflags & REG_EXPANDED) != 0) options |= PCRE_EXTENDED;
/* These ones only work on platforms where ints are wider than 16 bits */
if ((cflags & REG_DENDONLY) != 0) options |= PCRE_DOLLAR_ENDONLY;
if ((cflags & REG_ANCHORED) != 0) options |= PCRE_ANCHORED;
if ((cflags & REG_JAVASCPT) != 0) options |= PCRE_JAVASCRIPT_COMPAT;

if ((cflags & (REG_PCRE|REG_JAVASCPT)) == 0)
  {
  if ((cflags & REG_NEWLINE) == 0)  options |= PCRE_DOTALL | PCRE_DOLLAR_ENDONLY;
  if ((cflags & REG_NEWLINE) != 0)  options |= PCRE_NOT_EXCLUDES_NL;
  }

preg->re_impl = pcre_compile2(pattern, options, &errorcode, &errorptr,
  &erroffset, NULL);
preg->re_erroffset = erroffset;
if (dupez) free(dupez);

/* Safety: if the error code is too big for the translation vector (which
should not happen, but we all make mistakes), return REG_BADPAT. */

if (preg->re_impl == NULL)
  {
  return (errorcode < (int)(sizeof(eint)/sizeof(const int)))?
    eint[errorcode] : REG_BADPAT;
  }

(void)pcre_fullinfo((const pcre *)preg->re_impl, NULL, PCRE_INFO_CAPTURECOUNT,
  &re_nsub);
preg->re_nsub = (size_t)re_nsub;
preg->re_erroffset = (size_t)(-1);  /* No meaning after successful compile */
regcomp_jit(preg);
return 0;
}




/*************************************************
*              Match a regular expression        *
*************************************************/

/* Unfortunately, PCRE requires 3 ints of working space for each captured
substring, so we have to get and release working store instead of just using
the POSIX structures as was done in earlier releases when PCRE needed only 2
ints. However, if the number of possible capturing brackets is small, use a
block of store on the stack, to reduce the use of malloc/free. The threshold is
in a macro that can be changed at configure time.

If REG_NOSUB was specified at compile time, the PCRE_NO_AUTO_CAPTURE flag will
be set. When this is the case, the nmatch and pmatch arguments are ignored, and
the only result is yes/no/error. */

PCREPOSIX_EXP_DEFN int PCRE_CALL_CONVENTION
regexec(const regex_t *preg, const char *string, size_t nmatch,
  regmatch_t pmatch[], int eflags)
{
int rc, len;
size_t so, eo;
int options = 0;
int *ovector = NULL;
int small_ovector[POSIX_MALLOC_THRESHOLD * 3];
BOOL allocated_ovector = FALSE;
BOOL nosub;
const pcre *re_pcre = NULL;
const pcre_extra *re_pcre_extra = NULL;

if (preg == NULL || preg->re_impl == NULL) return REG_INVARG;
if ((eflags & REGEXEC_OPTIONS) != eflags) return REG_INVARG;

if (GET_MAGIC(preg->re_impl) == PCREPOSIX_JIT_MAGIC)
  {
  pcrx_t *pjit = (pcrx_t *)preg->re_impl;
  re_pcre_extra = pjit->re_pcre_extra;
  re_pcre = pjit->re_pcre;
  if (re_pcre == NULL) return REG_INVARG;
  }
else
  re_pcre = (pcre *)preg->re_impl;

nosub =
  (REAL_PCRE_OPTIONS(re_pcre) & PCRE_NO_AUTO_CAPTURE) != 0;

if ((eflags & REG_NOTBOL) != 0) options |= PCRE_NOTBOL;
if ((eflags & REG_NOTEOL) != 0) options |= PCRE_NOTEOL;
if ((eflags & REG_NOTEMPTY) != 0) options |= PCRE_NOTEMPTY;

/* When no string data is being returned, or no vector has been passed in which
to put it, ensure that nmatch is zero. Otherwise, ensure the vector for holding
the return data is large enough. */

if (nosub || pmatch == NULL) nmatch = 0;

else if (nmatch > 0)
  {
  if (nmatch <= POSIX_MALLOC_THRESHOLD)
    {
    ovector = &(small_ovector[0]);
    }
  else
    {
    if (nmatch > INT_MAX/(sizeof(int) * 3)) return REG_ESPACE;
    ovector = (int *)malloc(sizeof(int) * nmatch * 3);
    if (ovector == NULL) return REG_ESPACE;
    allocated_ovector = TRUE;
    }
  }

/* REG_STARTEND is a BSD extension, to allow for non-NUL-terminated strings.
The man page from OS X says "REG_STARTEND affects only the location of the
string, not how it is matched". That is why the "so" value is used to bump the
start location rather than being passed as a PCRE "starting offset". */

if ((eflags & REG_STARTEND) != 0)
  {
  if (pmatch == NULL) return REG_INVARG;
  if (pmatch[0].rm_so < 0 || pmatch[0].rm_eo < pmatch[0].rm_so) return REG_INVARG;
  so = (size_t)pmatch[0].rm_so;
  eo = (size_t)pmatch[0].rm_eo;
  if ((regoff_t)so != pmatch[0].rm_so || (regoff_t)eo != pmatch[0].rm_eo)
    return REG_INVARG;
  }
else
  {
  so = 0;
  eo = strlen(string);
  }

len = (int)(eo - so);
if ((size_t)len != (eo - so)) return REG_INVARG;
rc = pcre_exec(re_pcre, re_pcre_extra, string + so, len,
  0, options, ovector, (int)(nmatch * 3));

if (rc == 0) rc = (int)nmatch;    /* All captured slots were filled in */

/* Successful match */

if (rc >= 0)
  {
  size_t i;
  if (!nosub)
    {
    for (i = 0; i < (size_t)rc; i++)
      {
      pmatch[i].rm_so = (ovector[i*2] < 0)? (regoff_t)-1 : (regoff_t)ovector[i*2] + (regoff_t)so;
      pmatch[i].rm_eo = (ovector[i*2+1] < 0)? (regoff_t)-1: (regoff_t)ovector[i*2+1] + (regoff_t)so;
      }
    if (allocated_ovector) free(ovector);
    for (; i < nmatch; i++) pmatch[i].rm_so = pmatch[i].rm_eo = -1;
    }
  return 0;
  }

/* Unsuccessful match */

if (allocated_ovector) free(ovector);
switch(rc)
  {
/* ========================================================================== */
  /* These cases are never obeyed. This is a fudge that causes a compile-time
  error if the vector eint, which is indexed by compile-time error number, is
  not the correct length. It seems to be the only way to do such a check at
  compile time, as the sizeof() operator does not work in the C preprocessor.
  As all the PCRE_ERROR_xxx values are negative, we can use 0 and 1. */

  case 0:
  case (sizeof(eint)/sizeof(int) == ERRCOUNT):
  return REG_ASSERT;
/* ========================================================================== */

  case PCRE_ERROR_NOMATCH: return REG_NOMATCH;
  case PCRE_ERROR_NULL: return REG_INVARG;
  case PCRE_ERROR_BADOPTION: return REG_INVARG;
  case PCRE_ERROR_BADMAGIC: return REG_INVARG;
  case PCRE_ERROR_UNKNOWN_NODE: return REG_ASSERT;
  case PCRE_ERROR_NOMEMORY: return REG_ESPACE;
  case PCRE_ERROR_MATCHLIMIT: return REG_ESPACE;
  case PCRE_ERROR_BADUTF8: return REG_INVARG;
  case PCRE_ERROR_BADUTF8_OFFSET: return REG_INVARG;
  case PCRE_ERROR_BADMODE: return REG_INVARG;
  default: return REG_ASSERT;
  }
}

/* End of pcreposix.c */
